### ISSUE TITLE
nodejs edge case broken - x86 on apple silicon

### ISSUE THREAD

**Issue by liath** [wolfi-dev/os#37196]
This likely doesn't affect many people but I wanted to document it in case other people start googling for it. We're trialing chainguard and I noticed builds on my machine that use nodejs were hanging. Switching to standard node or ubuntu base images fixed this as did disabling Rosetta in docker which lead me to think chainguard/wolfi was doing something different. 

To reproduce, get an ARM mac device and set docker to use rosetta then:
```shell
# this will hang forever
docker run --rm -it --platform linux/amd64 cgr.dev/chainguard/node:latest-dev -e "console.log(10 % 4)"
```

As you can imagine, the `%` operator is used loads of places so this manifested for me in npm builds that would hang:
```shell
# this also hangs forever
$ npm i -ddd -g mimic-function
npm verbose cli /usr/bin/node /usr/bin/npm
...
npm silly idealTree buildDeps
npm silly fetch manifest mimic-function@*
npm silly packumentCache full:https://registry.npmjs.org/mimic-function cache-miss
```

Under gdb (against node@21.7.3) we can see the loop the process is stuck in:
![image](https://github.com/user-attachments/assets/2fb47b3c-4103-48d1-80d6-08b1852e243e)

It seems Rosetta doesn't translate `sahf` so the loop never exists. When compiled the same as official builds of nodejs are, they use stock gcc on ubuntu (so `march=x86-64 + mtune=generic`), we see the generated code avoids this instruction:
![image](https://github.com/user-attachments/assets/1138d0e1-1a22-42de-9483-3d027691c7c1)

I believe this all is caused by the tunes selected in [gcc.yaml](../blob/main/gcc.yaml), specifically [mtune=sapphirerapids](https://github.com/wolfi-dev/os/blob/main/gcc.yaml#L71) enables [a bunch of newer instructions](https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html#:~:text=%E2%80%98-,sapphirerapids,-%E2%80%99) which [Rosetta won't support](https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment#:~:text=Rosetta%20translates%20all%20x86_64%20instructions%2C%20but%20it%20doesn%E2%80%99t%20support%20the%20execution%20of%20some%20newer%20instruction%20sets%20and%20processor%20features%2C%20such%20as%20AVX%2C%20AVX2%2C%20and%20AVX512%20vector%20instructions.). Curiously `sahf` isn't listed in the Rosetta docs as not working but [others](https://github.com/Vita3K/Vita3K/pull/2038#issuecomment-1231936944) seem to be seeing issues with it.

**Potential workarounds:**
1. Disable Rosetta (you can disable it for one container with `EXPERIMENTAL_DOCKER_DESKTOP_FORCE_QEMU=1`)
2. Fetch [official](https://nodejs.org/en/download/prebuilt-binaries) nodejs build instead

**Comment by justinvreeland** [comment]
Thanks for the report. I appreciate the debugging you've done. It's good to know that SAHF is potentially buggy on rosetta that does look like it might be the culprit though I'm not sure why we're hitting it.

I have tried rebuilding node with `-mtune=generic` and the issue persists. I can also trace the issue with our node packages and rosetta to before we tuned to sapphire rapids.  Broadwell, our previous setting also has SAHF.

The link you provided doesn't seem to clarify why the poster thinks the instruction is buggy, both links to commits in the comment 404 for me and it's not clear to me what that PR adopted if anything to avoid SAHF instructions. SAHF is also a fairly old instruction as I understand it. So it seems surprising it wouldn't work well particularly since it was explicitly supported on [macOS intel cpus](https://developer.apple.com/documentation/xcode/writing-64-bit-intel-code-for-apple-platforms) so should've been well tested in rosetta.

Seems possible that when used in conjunction with other newer options we have set it causes problems but I have yet to figure out what that combination is.
